### PR TITLE
Add a global --flutter-from-path argument that overrides SDK selection

### PR DESCRIPTION
**Pull request by DanTup**
This promotes the `--flutter-from-path` flag to a global arg for all commands and uses it to locate the Flutter SDK before executing commands.

If set, we'll search for the SDK on PATH and fail if we don't find one.

If not set, we will infer an SDK from the one currently running the devtools_tool script (and again, fail if we're running from an SDK that doesn't have one).

When we call another instance of the tool, we bypass the shell script and instead execute it using Dart from the selected SDK (so that child processes use the selected SDK and we don't need to propagate the --flutter-from-path flag).

The update-flutter-sdk script was changed slightly to check whether the selected SDK is different to the tool/flutter-sdk one to decide whether to perform the extra update. This avoids some minor edge cases if:

- you have tool/flutter-sdk on PATH and pass the flag (it would update the Flutter SDK twice)
- you run devtools_tool from another SDK and don't pass the flag (it would not update the version being used to run the script)

Fixes https://github.com/flutter/devtools/issues/6867

### CODE REVIEW

File: tool/lib/commands/update_flutter_sdk.dart [hunk 1]
```diff
@@ -100,16 +90,28 @@ class UpdateFlutterSdkCommand extends Command {
       '${useCachedVersion ? 'from cache' : 'from upstream'}: $flutterTag ',
     );
 
-    if (updateFlutterFromPath) {
-      final sdk = FlutterSdk.current;
+    final sdk = FlutterSdk.current;
+    final flutterSdkDirName = 'flutter-sdk';
+    final toolSdkPath = path.join(repo.toolDirectoryPath, flutterSdkDirName);
+    final toolFlutterSdkDirectory = Directory(toolSdkPath);
```
Comment (kenzieschmoll) [inline tool/lib/commands/update_flutter_sdk.dart:96]: nit: add a getter `toolFlutterSdkPath` to the DevToolsRepo class

File: tool/lib/commands/update_flutter_sdk.dart [hunk 2]
```diff
@@ -33,22 +32,14 @@ const _useCacheFlag = 'use-cache';
...
 class UpdateFlutterSdkCommand extends Command {
   UpdateFlutterSdkCommand() {
-    argParser
-      ..addFlag(
-        _fromPathFlag,
-        abbr: 'p',
-        negatable: false,
-        help: 'Update the Flutter SDK that is on PATH (your local '
-            'flutter/flutter git checkout)',
-      )
```
Comment (kenzieschmoll) [inline tool/lib/commands/update_flutter_sdk.dart:43]: I think this should still be optional. We may not always want to wipe our local flutter checkout. Maybe we should change it to `--on-path` too.

File: tool/lib/devtools_command_runner.dart [hunk 3]
```diff
@@ -20,6 +22,9 @@ import 'commands/rollback.dart';
 import 'commands/update_dart_sdk_deps.dart';
 import 'commands/update_version.dart';
 
+const _flutterFromPathFlag = 'flutter-from-path';
+const _flutterFromPathAlias = 'from-path';
```
Comment (kenzieschmoll) [inline tool/lib/devtools_command_runner.dart:26]: nit: is the alias necessary? can we remove to simplify?

File: tool/lib/model.dart [hunk 4]
```diff
@@ -97,17 +97,26 @@ class DevToolsRepo {
...
+  /// The current located Flutter SDK.
   ///
   /// Tries to locate from the running Dart VM. If not found, will print a
   /// warning and use Flutter from PATH.
-  static final current = _findSdk();
+  static FlutterSdk get current {
+    if (_current == null) {
+      throw Exception(
+          'Cannot use FlutterSdk.current before SDK has been selected.'
+          'SDK selection is done by DevToolsCommandRunner.runCommand().');
```
Comment (kenzieschmoll) [inline tool/lib/model.dart:110]: nit: trailing comma

File: tool/lib/commands/update_flutter_sdk.dart (hunk 1 above)
Comment (DanTup) [reply tool/lib/commands/update_flutter_sdk.dart]: Done!

File: tool/lib/model.dart (hunk 4 above)
Comment (DanTup) [reply tool/lib/model.dart]: Done!

File: tool/lib/devtools_command_runner.dart (hunk 3 above)
Comment (kenzieschmoll) [reply tool/lib/devtools_command_runner.dart]: I think the 'p' abbreviation is sufficient for short hand 👍

File: tool/lib/commands/update_flutter_sdk.dart (hunk 2 above)
Comment (DanTup) [reply tool/lib/commands/update_flutter_sdk.dart:43]: > Uses the Flutter SDK from PATH to run the Processes started by the update-flutter-sdk command.
> vs
> Uses the `tool/flutter-sdk` to run the Processes started by the update-flutter-sdk command.

I guess my confusion is that it's not clear why we need this distinction. The two things you list that those commands do appears the same (update both Flutter SDKs). What's the purpose of running with the local tool/flutter-sdk SDK when updating the one from PATH?

I was trying to keep things simple by saying that basically you're always using one SDK or the other... Only one SDK exists in the world of the script, and you can choose the one from PATH or the one from the current running VM (which is essentially always tool/flutter-sdk because the shell scripts force that). Mixing them feels like it could cause confusion, because you need to know/remember which SDK is being used for what.


> Another way to summarize all this, if the intent of --flutter-from-path for all devtools_tool commands means "use my Flutter SDK on PATH for all processes and subprocesses started by this command", it feels strange for this flag to also mean "but when you run with this particular command update-flutter-sdk it actually does other things too".

I agree, and my original goal was for it not to do anything different at all, but rather just update whichever SDK was being used (and updated). However the original implementation seemed to update _both_ SDKs and I wasn't sure if it reasonable to change that. In summary:

- `--flutter-from-path` means always use the Flutter SDK from path for all spawned processes - Flutter or Dart, including devtools_tool. This occurs all the way down.
- `update-flutter-sdk` will always update the "current" SDK (based on that flag above), but additionally it will ensure "tool/flutter-sdk" is always up-to-date if it wasn't the one already being updated.

My preference would've been to remove that last part, (so update-flutter-sdk only ever updates one SDK - the one being run) because I think it's much easier to understand. If you want to update both SDKs, you could run `devtools_tool update-flutter-sdk && devtools_tool update-flutter-sdk --flutter-from-path`). I'm not sure if that addresses everything you wanted though?

File: tool/lib/commands/update_flutter_sdk.dart [hunk 5]
```diff
@@ -100,16 +98,23 @@ class UpdateFlutterSdkCommand extends Command {
...
-      final sdk = FlutterSdk.current;
+    final flutterSdkDirName = repo.sdkDirectoryName;
+    final toolSdkPath = repo.toolFlutterSdkPath;
 
-      log.stdout('Updating local flutter/flutter repository...');
+    // Check if the Flutter SDK we're using is not the local tool/flutter-sdk
+    // one and if so update that too.
+    //
+    // Check paths case-insensitively because of potential differences like
+    // windows drive letters.
```
Comment (kenzieschmoll) [inline tool/lib/commands/update_flutter_sdk.dart:108]: this comment seems outdated

File: tool/lib/devtools_command_runner.dart [hunk 6]
```diff
@@ -38,5 +42,25 @@ class DevToolsCommandRunner extends CommandRunner {
     addCommand(UpdateDevToolsVersionCommand());
     addCommand(UpdateFlutterSdkCommand());
     addCommand(UpdatePerfettoCommand());
+
+    argParser.addFlag(
+      _flutterFromPathFlag,
+      abbr: 'p',
+      negatable: false,
+      help: 'Use the Flutter SDK on PATH rather than the one being used to run '
+          'devtools_tool (usually tool/flutter-sdk)',
```
Comment (kenzieschmoll) [inline tool/lib/devtools_command_runner.dart:51]: nit this makes it sound like the tool/flutter-sdk is still being used to run `devtools_tool` even if the flutter from path flag is passed. Maybe reword: "Use the Flutter SDK on PATH to run devtools_tool commands and sub commands, instead of the Flutter SDK from tool/flutter-sdk which is used by default."

**Review by kenzieschmoll — APPROVED** [review]

File: tool/lib/commands/update_flutter_sdk.dart (hunk 5 above)
Comment (DanTup) [reply tool/lib/commands/update_flutter_sdk.dart]: Oops, updated!

File: tool/lib/devtools_command_runner.dart (hunk 6 above)
Comment (DanTup) [reply tool/lib/devtools_command_runner.dart]: Does my update make sense? Something that's slightly odd but I don't think matters enough to mention (because it makes things more verbose) is that this only applies to the spawned processes - the current process is using whatever the shell script decided to use (because we're already dunning a VM by the time we parse this flag).

I did consider once having the top level runner detect if this flag was set but it was running with a different SDK, and just re-run itself with the new one, but I'm not sure there would be any real benefit to it.

Comment (kenzieschmoll) [reply tool/lib/devtools_command_runner.dart]: If the only thing that could be run with a different SDK is the entrance to this script, then I think this is fine because all sub commands will respect the Flutter SDK we select.

Comment (DanTup) [reply tool/lib/devtools_command_runner.dart]: Yep, that's correct - as long as we don't add new code that accidentally runs things from path instead of using the current SDK, but as long as we go through `CliCommand.flutter` etc., then this should be hard to do accidentally.
